File: dynamic-programming/Fibonacci.java
/**
* @author 	Luca Cortinovis	luca1cortinovis@gmail.com
* @version 	1.0
* @since	2/11/2022
*
* You can find me here:
* GitHub	Luca Cortinovis(NeptuneLuke)
*/


import java.util.Scanner;
import java.util.Arrays;	

public class Fibonacci {

	
	public static void main(String args[]) {
		
		Scanner keyboard = new Scanner(System.in);
	
		System.out.println("Insert the nth number of Fibonacci: ");
		int n = keyboard.nextInt();
		
		int fib [] = fibonacci_iter(n);
		System.out.println("Fibonacci sequence: ");
		System.out.println(Arrays.toString(fib));
		
		int fib_num = fibonacci_rec(n);
		System.out.println("Fibonacci recursive: " + fib_num);	
	}
	
	
	/* This is an iterative algorithm of Fibonacci sequence
	 * Time complexity: O(n)
	 */
	public static int[] fibonacci_iter(int n) {
	
		int A[] = new int [n+1];
		A[0] = 0;
		A[1] = 1;
		for(int i=2; i<=n; i++)
			A[i] = A[i-1] + A[i-2];
		
		return A;
	}
	
	
	/* This is the recursive algorithm for nth Fibonacci number
	 * Time complexity: O( ((1-sqrt(5))/2 )^n ) exponential
	 */
	public static int fibonacci_rec(int n) {
		
		if(n<=1)
			return n;
		else
			return fibonacci_rec(n-1) + fibonacci_rec(n-2);
	}
	
}

File: dynamic-programming/fibonacci.cpp
/**
	* @author 	Luca Cortinovis	luca1cortinovis@gmail.com
	* @since	28/10/2023
	*
	* You can find me here:
	* GitHub	Luca Cortinovis (NeptuneLuke)
*/

#include <iostream>
#include <string>
#include <vector>


using
std::string,
std::vector,
std::cout,
std::cin;


vector<int> fibonacci_iter(int n);
int fibonacci_rec(int n);


int main(void) {

	cout << "Insert the nth number of Fibonacci: \n";
	int n = 0;
	cin >> n;
	cout << "\n\n";


	// Fibonacci iterative
	vector<int> fib = fibonacci_iter(n);
	cout << "Fibonacci sequence: ";
	for (auto elem : fib) {
		cout << elem << " ";
	}
	cout << "\n\n";


	// Fibonacci recursive
	int fib_num = fibonacci_rec(n);
	cout << "Fibonacci recursive: " << fib_num;


	return 0;
}


vector<int> fibonacci_iter(int n) {

	vector<int> fib(n + 1);
	fib[0] = 0;
	fib[1] = 1;

	for (int i = 2; i <= n; i++)
		fib[i] = fib[i - 1] + fib[i - 2];

	return fib;
}

int fibonacci_rec(int n) {

	if (n <= 1)
		return n;
	else
		return fibonacci_rec(n - 1) + fibonacci_rec(n - 2);
}